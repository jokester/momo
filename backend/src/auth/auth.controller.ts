import { BadRequestException, Body, Controller, Get, Post, Req, Request, UseGuards } from '@nestjs/common';
import { AuthService } from './auth.service';
import { UserEntity } from '../entity/user.entity';
import { LocalStrategyGuard } from './local.strategy';
import { getDebugLogger } from '../util/get-debug-logger';
import { GoogleOAuthService } from './google-oauth.service';
import { UserService } from '../user/user.service';

const logger = getDebugLogger(__filename);

@Controller('auth')
export class AuthController {
  constructor(
    private readonly authService: AuthService,
    private readonly googleOAuthService: GoogleOAuthService,
    private readonly userService: UserService,
  ) {}

  @UseGuards(LocalStrategyGuard)
  @Post('email/signin')
  async emailSignIn(@Request() req: Request & { user?: UserEntity }): Promise<UserEntity | null> {
    logger('got email user: %o', req.user);
    return req.user || null;
  }

  @Get('google/oauth' /* ?code=callback_code... */)
  async startGoogleOAuth(@Request() req: Request & { user?: UserEntity }) {
    logger('got google/oauth user: %o', req.user);
    return null;
  }

  @Get('ping')
  async pingDb() {
    await this.userService.now();
    return null;
  }

  @Post('oauth/google')
  async doGoogleOAuth(@Request() req: Request, @Body() payload2: { code: string; callbackUri: string }) {
    // const payload: { code: string; callbackUri: string } = await req.json();

    const payload = payload2;
    if (payload && payload.code) {
      const x = await this.googleOAuthService.auth(payload.code);

      logger('got auth', x);
      return x;
    }
    throw new BadRequestException();
  }
}
