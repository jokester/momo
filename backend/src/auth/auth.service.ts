import { Injectable } from '@nestjs/common';
import { getDebugLogger } from '../util/get-debug-logger';

@Injectable()
export class AuthService {
  constructor() {
    getDebugLogger(__filename)('authService created');
  }

  touch() {
    getDebugLogger(__filename)('authService touched');
  }
}
