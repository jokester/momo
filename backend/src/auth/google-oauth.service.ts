import { google } from 'googleapis';
import { Injectable } from '@nestjs/common';
import { ConfigService } from '@nestjs/config';
import { getDebugLogger } from '../util/get-debug-logger';
import { oauth2 } from 'googleapis/build/src/apis/oauth2';

const logger = getDebugLogger(__filename);

@Injectable()
export class GoogleOAuthService {
  constructor(private readonly confService: ConfigService) {}

  async auth(code: string) {
    const client = this.createAuthClient();
    logger('client=%O', client);
    logger('code=', code);

    const token = await client.getToken(code);
    client.setCredentials(token.tokens);
    logger('token=%O', token);

    const x = await oauth2({ version: 'v2', auth: client });
    const got = await x.userinfo.get();
    return got.data;
  }

  private createAuthClient() {
    return new google.auth.OAuth2({
      clientId: this.confService.get('OAUTH_GOOGLE_CLIENT_ID'),
      clientSecret: this.confService.get('OAUTH_GOOGLE_CLIENT_SECRET'),
      redirectUri: this.confService.get('OAUTH_GOOGLE_REDIRECT_URI'),
    });
  }
}
