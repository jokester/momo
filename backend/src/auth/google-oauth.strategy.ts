import { AuthGuard, PassportStrategy } from '@nestjs/passport';
import { Profile, Strategy } from 'passport-google-oauth20';
import { UserEntity } from '../entity/user.entity';
import { Injectable, InternalServerErrorException } from '@nestjs/common';
import { getDebugLogger } from '../util/get-debug-logger';

@Injectable()
export class GoogleOauthStrategy extends PassportStrategy(Strategy, 'google') {
  constructor() {
    super({
      clientID: 'FILL ME',
      clientSecret: 'FILL_ME',
      scope: ['profile', 'email'],
      callbackURL: 'FILL_ME', // have to match the one retrieved code?
    });
  }

  async validate(accessToken: string, refreshToken: string, profile: Profile): Promise<UserEntity> {
    getDebugLogger(__filename)('%s / %s / %O', accessToken, refreshToken, profile);

    throw new InternalServerErrorException('something is wrong');
  }
}

export const GoogleOAuthGuard = AuthGuard('google');
