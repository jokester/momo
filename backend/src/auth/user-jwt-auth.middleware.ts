import {
  BadRequestException,
  createParamDecorator,
  ExecutionContext,
  Injectable,
  NestMiddleware,
  UnauthorizedException,
} from '@nestjs/common';
import { Request, Response } from 'express';
import { UserService } from '../user/user.service';
import { getDebugLogger } from '../util/get-debug-logger';
import { getRightOrThrow } from '../util/fpts-getter';
import { UserAccount } from '../db/entities/user-account';
import { ErrorCodeEnum } from '../const/error-code';

const logger = getDebugLogger(__filename);

/**
 * @class UserJwtAuthMiddleware
 * @desc attach authorized user id to {@link Request} object
 */
@Injectable()
export class UserJwtAuthMiddleware implements NestMiddleware {
  constructor(private readonly userService: UserService) {}

  async use(req: Request, res: Response, next: () => void): Promise<void> {
    const authHeader = req.header('Authorization');
    logger('running', authHeader);

    if (authHeader) {
      const [_whatever, jwtToken] = /^Bearer ([^ ]*)$/.exec(authHeader) || [];

      if (!jwtToken) {
        throw new BadRequestException();
      }
      const user = getRightOrThrow(
        await this.userService.findUserWithJwtToken(jwtToken),
        (l) => new UnauthorizedException('auth required', l),
      );
      (req as any).authedUser = user;
    }

    next();
  }
}

export const AuthedUser = createParamDecorator((data: unknown, ctx: ExecutionContext) => {
  const { authedUser } = ctx.switchToHttp().getRequest<{ authedUser?: UserAccount }>();
  if (!authedUser) {
    throw new UnauthorizedException('auth required', ErrorCodeEnum.notAuthenticated);
  }
  return authedUser;
});
