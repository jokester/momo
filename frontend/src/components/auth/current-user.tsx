import React from 'react';
import { UserProfileDto } from '../../services/api-generated/models';
import { useSingletons } from '../../internal/app-context';
import gravatarUrl from 'gravatar-url';
import { Button } from '@blueprintjs/core';

export const CurrentUser: React.FC<{ user: UserProfileDto }> = ({ user }) => {
  const { auth } = useSingletons();
  return (
    user && (
      <div>
        <p>
          <img className="h-64 w-64" src={gravatarUrl(user.email || 'user@example.com', { size: 320 })} />
          {user.email}
        </p>
        <p>
          用户id: <span className="monospace">{user.userId}</span>
        </p>
        <br />
        <Button onClick={() => auth.signOut()}>退出登录</Button>
      </div>
    )
  );
};
