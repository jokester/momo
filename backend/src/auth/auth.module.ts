import { Module } from '@nestjs/common';
import { AuthController } from './auth.controller';
import { DiscordOAuth, GoogleOAuth } from './oauth-client.provider';
import { DiscordOAuthService } from './discord-oauth.service';
import { GoogleOAuthService } from './google-oauth.service';
import { UserModule } from '../user/user.module';

@Module({
  imports: [UserModule],
  controllers: [AuthController],
  providers: [DiscordOAuth.Provider, DiscordOAuthService, GoogleOAuth.Provider, GoogleOAuthService],
})
export class AuthModule {}
