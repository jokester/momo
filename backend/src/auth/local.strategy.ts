import { AuthGuard, PassportStrategy } from '@nestjs/passport';
import { Strategy } from 'passport-local';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { AuthService } from './auth.service';
import { getDebugLogger } from '../util/get-debug-logger';
import { UserEntity } from '../entity/user.entity';

@Injectable()
export class LocalStrategy extends PassportStrategy(Strategy) {
  constructor(private readonly authService: AuthService) {
    super();
    getDebugLogger(__filename)('gotAuthService: %o', authService);
  }

  async validate(username: string, password: string): Promise<UserEntity> {
    const user = await this.authService.authUserWithPassword(username, password);

    if (user.isSome()) {
      return user.value;
    }
    throw new UnauthorizedException();
  }
}

export const LocalStrategyGuard = AuthGuard('local');
