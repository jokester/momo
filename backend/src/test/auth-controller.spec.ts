import { AuthController } from '../user/auth.controller';
import { buildTesteeAppBundle } from './test-app-factory';
import { INestApplication } from '@nestjs/common';
import { MockData, TestDeps } from './test-deps';
import request from 'supertest';
import { JwtService } from '@nestjs/jwt';
import { DiscordOAuth } from '../user/oauth-client.provider';

describe(AuthController, () => {
  const testee = buildTesteeAppBundle();
  let app: INestApplication;
  let jwtService: JwtService;
  let discordOAuthClient: DiscordOAuth.Client;
  let authController: AuthController;

  beforeAll(async () => {
    await testee.bundledBeforeAll();
    app = testee.testBundle.app;
    jwtService = testee.testBundle.jwtService;
    discordOAuthClient = testee.testBundle.discordOAuthClient;
    authController = app.get(AuthController);
  });
  afterAll(testee.bundledAfterAll);

  beforeEach(TestDeps.resetTestDB);

  describe('POST /auth/oauth/discord', () => {
    describe('when discord allows oauth', () => {
      beforeEach(() => {
        jest
          // @ts-ignore
          .spyOn(discordOAuthClient, 'oauthCallback')
          .mockResolvedValue(MockData.discordOAuthTokenValid);

        jest
          // @ts-ignore
          .spyOn(discordOAuthClient, 'userinfo')
          .mockResolvedValue(MockData.discordOAuthUserInfoValid);
      });

      it('creates user at controller', async () => {
        const x = await authController.doDiscordOAuth(MockData.oauthRequest);
      });

      it('creates UserAccount and OAuthAccount when none existed', async () => {
        const res = await request(app.getHttpServer())
          .post('/auth/oauth/discord')
          .send(MockData.oauthRequest)
          .expect(201);

        const jwtToken = JSON.parse(res.text).jwtToken;
        await jwtService.verifyAsync(jwtToken);
      });
    });
  });
});
