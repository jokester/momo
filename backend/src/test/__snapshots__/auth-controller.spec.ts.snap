// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthController POST /auth/email/signup returns 201 + user on success: POST /auth/email/signup 201 1`] = `
Object {
  "email": "a@b.com",
  "userId": "nanoid-5",
}
`;

exports[`AuthController POST /auth/oauth/google when google returns unverified user profile return 400 on auth error: auth 400 res 1`] = `
Object {
  "error": "email must be verified",
  "message": "auth failed",
  "statusCode": 400,
}
`;
